pkg/util/kube/pod.go

```diff
@@ -110,3 +110,37 @@ func EnsureDeletePod(ctx context.Context, podGetter corev1client.CoreV1Interface
 
 	return nil
 }
+
+// IsPodInAbnormalState checks if the pod is in an abnormal state
+// It could not cover all the cases but we could add more cases in the future
+func IsPodInAbnormalState(pod *corev1api.Pod, log logrus.FieldLogger) bool {
+	// Check the Phase field
+	if pod.Status.Phase == corev1api.PodFailed || pod.Status.Phase == corev1api.PodUnknown {
+		log.Debugf("Pod is in abnormal state %s", pod.Status.Phase)
+		return true
+	} else if pod.Status.Phase == corev1api.PodPending {
+		// Check the conditions for Pending reason to see if it's unschedulable
+		for _, condition := range pod.Status.Conditions {
+			if condition.Reason == corev1api.PodReasonUnschedulable {
+				log.Debug("Pod is in unschedulable state")
+				return true
+			}
+		}
+	}
+
+	// Check the Status field
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		// Check for container status not ready
+		if !containerStatus.Ready && containerStatus.State.Running != nil { // Readiness probe failure
```

Comment:
I think it is also risky to make this assert --- when a pod is in pending status, if one of the container is not ready, then we say the pod is not in normal status.
Pod's startup goes through multiple operations, for example, it first request volumes from storage, unless the storage makes the volumes ready, it won't start the container. While requesting volumes takes some time. So it is likely that we see a pod is in pending status and one or more of its container is not in running status. However, it is not actually an abnormal status.

Reply:
it means that one specific container is in running status but the service readiness probe failed.
so if the container is running it has already stared up.

Reply:
Readiness probe failure can be covered in some cases.
It depends on the readiness probe check condition. If it relies on some external services, when the external services are available, the readiness probe will succeed.

Reply:
When the readiness probe fails for once, could it recover later?
I think the readiness probe is done by kubelet according to the probe command defined in the container, if the probe command is not defined, what is the behavior? Does it still do the probe?

Reply:
It's possible that the probe command fails at first, then the command succeeds after some time.
If the probe command is not defined, then there will be no readiness check. After the pod enters the Running state, the pod also enters the Ready state.

Reply:
> When the readiness probe fails for once, could it recover later?

from my memory, readiness probe command is exected periodically and ready status updated accordingly. most commonly curling an `endpoint:port/path` to check server is ready to serve sites.

Reply:
I removed the readiness check first for a later more specific scenario we could add it back